flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,301 @@
+     * plane through the center of the box keeping the n-th coordinate fixed,
+     * then shifting it in the n-th direction up and down
+     * and recursively applying partitionBox to the two shifted (n-1) dimensional planes.
+     *
+     * @param center
+     * @param width
+     * @return
+     *
+     */
+    def partitionBox(center: Vector, width: Vector): Seq[Vector] = {
+
+      def partitionHelper(box: Seq[Vector], dim: Int): Seq[Vector] = {
+        if (dim >= width.size) {
+          box
+        } else {
+          val newBox = box.flatMap {
+            vector =>
+              val (up, down) = (vector.copy, vector)
+              up.update(dim, up(dim) - width(dim) / 4)
+              down.update(dim, down(dim) + width(dim) / 4)
+
+              Seq(up,down)
+          }
+          partitionHelper(newBox, dim + 1)
+        }
+      }
+      partitionHelper(Seq(center), 0)
+    }
+  }
+
+
+  val root = new Node( ((minVec.asBreeze + maxVec.asBreeze)*0.5).fromBreeze,
+    (maxVec.asBreeze - minVec.asBreeze).fromBreeze, null)
+
+    /**
+     *   simple printing of tree for testing/debugging
+     */
+  def printTree(){
+    printTreeRecur(root)
+  }
+
+  def printTreeRecur(n:Node){
+    if(n.children != null) {
+      for (c <- n.children){
+        printTreeRecur(c)
+      }
+    }else{
+      println("printing tree: n.objects " + n.objects)
+    }
+  }
+
+  /**
+   * Recursively adds an object to the tree
+   * @param obj
+   */
+  def insert(obj:Vector){
+    insertRecur(obj,root)
+  }
+
+  private def insertRecur(obj:Vector,n:Node) {
+    if(n.children==null) {
+      if(n.objects.length < maxPerBox )
+      {
+        n.objects += obj
+      }
+
+      else{
+        n.makeChildren()  ///make children nodes; place objects into them and clear node.objects
+        for (o <- n.objects){
+          insertRecur(o, n.children(n.whichChild(o)))
+        }
+        n.objects.clear()
+        insertRecur(obj, n.children(n.whichChild(obj)))
+      }
+    } else{
+      insertRecur(obj, n.children(n.whichChild(obj)))
+    }
+  }
+
+  /** Following methods are used to zoom in on a region near a test point for a fast KNN query.
+    *
+    * This capability is used in the KNN query to find k "near" neighbors n_1,...,n_k, from
+    * which one computes the max distance D_s to obj.  D_s is then used during the
+    * kNN query to find all points within a radius D_s of obj using searchNeighbors.
+    * To find the "near" neighbors, a min-heap is defined on the leaf nodes of the quadtree.
+    * The priority of a leaf node is an appropriate notion of the distance between the test
+    * point and the node, which is defined by minDist(obj),
+   *
+   */
+  private def subOne(tuple: (Double,Node)) = tuple._1
+
+  def searchNeighborsSiblingQueue(obj:Vector):ListBuffer[Vector] = {
+    var ret = new ListBuffer[Vector]
+    if (root.children == null) {   // edge case when the main box has not been partitioned at all
+      root.objects
```

Comment:
You could also return a `List[Vector]` and then simply call `toList`. This will copy the `ListBuffer` instance only when it is modified afterwards.